selfdrive/test/test_updated.py

```diff
@@ -0,0 +1,205 @@
+import subprocess
+
+from common.basedir import BASEDIR
+from common.params import Params
+
+
+class TestUpdater(unittest.TestCase):
+
+  def setUp(self):
+    self.updated_proc = None
+
+    self.tmp_dir = tempfile.TemporaryDirectory()
+    org_dir = os.path.join(self.tmp_dir.name, "commaai")
+
+    self.basedir = os.path.join(org_dir, "openpilot")
+    self.git_remote_dir = os.path.join(org_dir, "openpilot_remote")
+    self.staging_dir = os.path.join(org_dir, "safe_staging")
+    for d in [org_dir, self.basedir, self.git_remote_dir, self.staging_dir]:
+      os.mkdir(d)
+
+    # setup local submodule remotes
+    submodules = subprocess.check_output("git submodule --quiet foreach 'echo $name'",
+                                         shell=True, cwd=BASEDIR, encoding='utf8').split()
+    for s in submodules:
+      sub_path = os.path.join(org_dir, s.split("_repo")[0])
+      self._run(f"git clone {s} {sub_path}.git", cwd=BASEDIR)
+
+    # setup two git repos, a remote and one we'll run updated in
+    self._run([
+      f"git clone {BASEDIR} {self.git_remote_dir}",
+      f"git clone {self.git_remote_dir} {self.basedir}",
+      f"cd {self.basedir} && git submodule init && git submodule update",
+      f"cd {self.basedir} && scons -j4 cereal"
+    ])
+
+    self.params = Params(db=os.path.join(self.basedir, "persist/params"))
+    self.params.clear_all()
+    os.sync()
+
+  def tearDown(self):
+    try:
+      if self.updated_proc is not None:
+        self.updated_proc.terminate()
+        self.updated_proc.wait(30)
+    except Exception as e:
+      print(e)
+    self.tmp_dir.cleanup()
+
+  def _run(self, cmd, cwd=None):
+    if not isinstance(cmd, list):
+      cmd = (cmd,)
+
+    for c in cmd:
+      subprocess.check_output(c, cwd=cwd, shell=True)
+
+  def _get_updated_proc(self):
+    os.environ["PYTHONPATH"] = self.basedir
+    os.environ["UPDATER_TESTING"] = "1"
+    os.environ["UPDATER_LOCK_FILE"] = os.path.join(self.tmp_dir.name, "updater.lock")
+    os.environ["UPDATER_STAGING_ROOT"] = self.staging_dir
+    updated_path = os.path.join(self.basedir, "selfdrive/updated.py")
+    return subprocess.Popen(updated_path, env=os.environ)
+
+  def _start_updater(self, offroad=True):
+    self.params.put("IsOffroad", "1" if offroad else "0")
+    self.updated_proc = self._get_updated_proc()
+
+  def _update_now(self):
+    self.updated_proc.send_signal(signal.SIGHUP)
+
+  def _wait_for_update(self, timeout=30, clear_param=False):
+    if clear_param:
+      self.params.delete("LastUpdateTime")
+
+    self._update_now()
+    start_time = time.monotonic()
+    while self.params.get("LastUpdateTime") is None:
+      if time.monotonic() - start_time > timeout:
+        raise Exception("timed out waiting for update to complate")
+      time.sleep(0.05)
+
+  def _make_commit(self):
+    self._run([
+      "git config user.email tester@testing.com",
+      "git config user.name Testy Tester",
+      "git commit --allow-empty -m 'an update'",
```

Comment:
Maybe don't make an empty commit? Test for:
 - Creating a new file
 - Deleting an existing file
 - Making a change to a file

And verify that all three basic operations work correctly?

Reply:
those operations and more are now done in the commit